core-processor/src/main/java/io/micronaut/inject/writer/BeanDefinitionWriter.java

```diff
@@ -2369,24 +2396,109 @@ private StatementDef addInnerConfigurationMethod() {
     }
 
     private StatementDef addGetExposedTypes() {
-        if (annotationMetadata.hasDeclaredAnnotation(Bean.class.getName())) {
-            final String[] exposedTypes = annotationMetadata.stringValues(Bean.class.getName(), "typed");
-            if (exposedTypes.length > 0) {
-                FieldDef exposedTypesField = FieldDef.builder(FIELD_EXPOSED_TYPES, Set.class)
-                    .addModifiers(Modifier.PRIVATE, Modifier.FINAL, Modifier.STATIC)
-                    .build();
+        AnnotationMetadata producingAnnotationMetadata;
+        if (beanProducingElement instanceof MethodElement methodElement) {
+            producingAnnotationMetadata = methodElement.getMethodAnnotationMetadata();
+        } else {
+            producingAnnotationMetadata = annotationMetadata;
+        }
+        String[] exposedTypes = producingAnnotationMetadata.stringValues(Bean.class.getName(), "typed");
+        Set<String> exposedTypeNames;
+        if (exposedTypes.length != 0) {
+            exposedTypeNames = Set.of(exposedTypes);
+        } else {
+            exposedTypeNames = new HashSet<>();
```

Comment:
can you make sure that the order of the exposed interfaces it always written out in the same order otherwise it might break incremental compilation. This may need to be a TreeSet sorted by interface name